### DISCUSSION TITLE
How to use async/await in iter map?

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/tokio#4346] [Q&A] [answered]
```rust
    let ids = vec![1, 2, 3];
    let data = ids
        .into_iter()
        .map(|x| async {
            // httpGetData(id).await
        })
        .collect();
```

**Comment by Darksonn** [comment, marked as answer]
The way to do this is to use a `Stream` rather than an `Iterator`. Streams have a method similar to `map` called `then` which lets you perform async operations.
```Rust
use futures::stream::StreamExt;

async fn http_get_data(id: i32) -> String {
    id.to_string()
}

#[tokio::main]
async fn main() {
    let ids = vec![1, 2, 3];
    let data: Vec<_> = futures::stream::iter(ids)
        .then(|id| async move {
            http_get_data(id).await
        })
        .collect()
        .await;

    println!("{:?}", data);
}
```
Note that the async block is not actually necessary in this particular example:
```Rust
let data: Vec<_> = futures::stream::iter(ids)
    .then(|id| http_get_data(id))
    .collect()
    .await;
```

**Reply by lz1998** [reply]
> The way to do this is to use a `Stream` rather than an `Iterator`. Streams have a method similar to `map` called `then` which lets you perform async operations.
> 
> ```rust
> use futures::stream::StreamExt;
> 
> async fn http_get_data(id: i32) -> String {
>     id.to_string()
> }
> 
> #[tokio::main]
> async fn main() {
>     let ids = vec![1, 2, 3];
>     let data: Vec<_> = futures::stream::iter(ids)
>         .then(|id| async move {
>             http_get_data(id).await
>         })
>         .collect()
>         .await;
> 
>     println!("{:?}", data);
> }
> ```
> 
> Note that the async block is not actually necessary in this particular example:
> 
> ```rust
> let data: Vec<_> = futures::stream::iter(ids)
>     .then(|id| http_get_data(id))
>     .collect()
>     .await;
> ```

Then how to execute requests at the same time with a concurrent limit?

**Reply by Darksonn** [reply]
Check out [this blog post][1] for that.

[1]: https://gendignoux.com/blog/2021/04/01/rust-async-streams-futures-part1.html
